File: BuddyApp/app/src/main/java/com/example/buddyapp/groupchat/CreateGroup.java
package com.example.buddyapp.groupchat;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.buddyapp.AllUserMember;
import com.example.buddyapp.R;
import com.firebase.ui.database.FirebaseRecyclerAdapter;
import com.firebase.ui.database.FirebaseRecyclerOptions;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import java.util.ArrayList;
import java.util.HashSet;

public class CreateGroup extends AppCompatActivity {

    DatabaseReference profileRef,documentReferenceGroup;
    HashSet<String> ids;
    FirebaseRecyclerAdapter<AllUserMember, UserSelectionViewHolder> firebaseRecyclerAdapter;
    FirebaseDatabase database = FirebaseDatabase.getInstance("https://buddyapp-5a091-default-rtdb.europe-west1.firebasedatabase.app/");
    RecyclerView recyclerView;
    EditText groupName;
    Boolean startChecking = false;
    Button createGroupbtn;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        documentReferenceGroup =database.getReference("groups");
        ids = new HashSet<String>();
        setContentView(R.layout.activity_create_group);
        recyclerView = findViewById(R.id.rv_users_for_group);
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(CreateGroup.this));
        profileRef = database.getReference("All Users");
        groupName = findViewById(R.id.groupname_create);
        createGroupbtn = findViewById(R.id.creategroup_btn);
        createGroupbtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                startChecking = true;
                startCreatingGroup();
            }
        });
    }

    protected void onStart() {
        super.onStart();
        FirebaseRecyclerOptions<AllUserMember> options =
                new FirebaseRecyclerOptions.Builder<AllUserMember>()
                        .setQuery(profileRef, AllUserMember.class)
                        .build();
        firebaseRecyclerAdapter =
                new FirebaseRecyclerAdapter<AllUserMember, UserSelectionViewHolder>(options) {
                    @Override
                    protected void onBindViewHolder(@NonNull UserSelectionViewHolder holder, int position, @NonNull AllUserMember model) {
                        final String postkey = getRef(position).getKey();
                        holder.setUserSelection(getApplication(), model.getName(), model.getUid(), model.getProf(), model.getUrl());
                        String name = getItem(position).getName();
                        String url = getItem(position).getUrl();
                        String uid = getItem(position).getUid();
                    }

                    @NonNull
                    @Override
                    public UserSelectionViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
                        View view = LayoutInflater.from(parent.getContext())
                                .inflate(R.layout.user_selection_for_group, parent, false);
                        return new UserSelectionViewHolder(view);
                    }
                };
        firebaseRecyclerAdapter.startListening();
        recyclerView.setAdapter(firebaseRecyclerAdapter);
    }

    protected void startCreatingGroup() {
        ids.clear();
        FirebaseUser user = FirebaseAuth.getInstance().getCurrentUser();
        ids.add(user.getUid());
        for (int i = 0; i < firebaseRecyclerAdapter.getItemCount(); i++) {
            UserSelectionViewHolder viewHolder = (UserSelectionViewHolder) recyclerView.findViewHolderForAdapterPosition(i);
            if (viewHolder != null) {
                if (viewHolder.checkBox.isChecked()) {
                    // This user is selected, you can do whatever you want here
                    // For example, you can add the user's ID to your group
                    AllUserMember selectedUser = firebaseRecyclerAdapter.getItem(i);
                    if (selectedUser != null) {
                        ids.add(selectedUser.getUid()); // Assuming getUid() returns the user ID
                    }
                }
            }
        }
        if(ids.size()>2 && !(groupName.getText().equals(null))){
            PojoGroup pojoGroup=new PojoGroup(new ArrayList<>(ids),user.getUid(),groupName.getText().toString());
            String keychild=documentReferenceGroup.push().getKey();
            documentReferenceGroup.child(keychild).setValue(pojoGroup);
            Intent intent=new Intent(CreateGroup.this, GroupRoom.class);
            startActivity(intent);
        }
        else{
            Toast.makeText(CreateGroup.this,"Select atleast 2 member",Toast.LENGTH_LONG).show();
        }
    }
}

File: RahilChatApplication/app/src/main/java/com/example/rahilchatapplication/CreateGroup.kt
package com.example.rahilchatapplication

import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase

class CreateGroup : AppCompatActivity(){
    private lateinit var etGroupName: EditText
    private lateinit var btnCreateGroup: Button
    private lateinit var btnListGroups: Button
    private lateinit var mDbRef: DatabaseReference
    private lateinit var mAuth: FirebaseAuth

    @RequiresApi(Build.VERSION_CODES.M)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.create_group)

        etGroupName = findViewById(R.id.etGroupName)
        btnCreateGroup = findViewById(R.id.btnCreateGroup)
        btnListGroups = findViewById(R.id.btnListGroups)
        mDbRef = FirebaseDatabase.getInstance().reference

        btnCreateGroup.setOnClickListener {
            val groupName = etGroupName.text.toString().trim()
            if (groupName.isNotEmpty()) {
                addGroupToFirebase(groupName)
            } else {
                showToast("Please enter a group name.")
            }
        }

        btnListGroups.setOnClickListener {
            val intent = Intent(this, GroupList::class.java)
            finish()
            startActivity(intent)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    private fun addGroupToFirebase(groupName: String) {
        val groupsRef = mDbRef.child("groups")
        val groupKey = groupsRef.push().key

        groupKey?.let {
            groupsRef.child(it).child("name").setValue(groupName)
                .addOnSuccessListener {
                    showToast("Group added successfully to Firebase Database.")
                }
                .addOnFailureListener { exception ->
                    showToast("Failed to add group to Firebase Database: $exception")
                }
        } ?: run {
            showToast("Failed to get group key.")
        }
    }

    private fun showToast(message: String) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
    }
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if(item.itemId == R.id.create_group) {
            val intent = Intent(this, CreateGroup::class.java)
            finish()
            startActivity(intent)
            return true;
        } else if(item.itemId == R.id.list_groups) {
            val intent = Intent(this, GroupList::class.java)
            finish()
            startActivity(intent)
            return true;
        }else if(item.itemId == R.id.chat_list) {
            val intent = Intent(this, MainActivity::class.java)
            finish()
            startActivity(intent)
            return true;
        }

        return true;
    }

}
